kubespawner/spawner.py

```diff
@@ -1911,6 +1899,40 @@ def _start(self):
             raise Exception(
                 'Can not create user pod %s already exists & could not be deleted' % self.pod_name)
 
+
+        if self.cert_paths:
+            yield exponential_backoff(
+                lambda: self.is_pod_creating(self.pod_reflector.pods.get(self.pod_name, None)),
+                'pod/%s does not exist!' % (self.pod_name),
+            )
+
+            pod = self.pod_reflector.pods[self.pod_name]
+            owner = make_owner_reference(self.pod_name, pod.metadata.uid)
+            try:
+                yield self.asynchronize(
+                    self.api.create_namespaced_secret,
+                    namespace=self.namespace,
+                    body=self.get_secret_manifest(owner)
+                )
+            except ApiException as e:
+                self.log.info(e)
+                if e.status == 409:
+                    self.log.info("Secret " + self.secret_name + " already exists, so did not create new secret.")
+                else:
+                    raise
```

Comment:
Are user certificates are recreated by JupyterHub? This could cause issues. The service should be deleted when the past pod has been deleted, so I think it won't happen, but if this occurs, this will probably cause a proper failure for the user.

This should not be only self.log.info, or have a crash loop backoff attempt process.

Reply:
The self.log.info was a mistake.  As for the other stuff I'm not following what you saying could you please explain differently.  Whenever the pod is deleted both the secret and the service will be garbage collected.

Reply:
It regards implementing the same kind of logic that can be found just above for attempting to create the pod multiple times and only accept failure after a while.

```python
        # If we run into a 409 Conflict error, it means a pod with the
        # same name already exists. We stop it, wait for it to stop, and
        # try again. We try 4 times, and if it still fails we give up.
        # FIXME: Have better / cleaner retry logic!
        retry_times = 4
        pod = yield self.get_pod_manifest()
        if self.modify_pod_hook:
            pod = yield gen.maybe_future(self.modify_pod_hook(self, pod))
        for i in range(retry_times):
            try:
                yield self.asynchronize(
                    self.api.create_namespaced_pod,
                    self.namespace,
                    pod,
                )
                break
            except ApiException as e:
                if e.status != 409:
                    # We only want to handle 409 conflict errors
                    self.log.exception("Failed for %s", pod.to_str())
                    raise
                self.log.info('Found existing pod %s, attempting to kill', self.pod_name)
                # TODO: this should show up in events
                yield self.stop(True)
                self.log.info('Killed pod %s, will try starting singleuser pod again', self.pod_name)
        else:
            raise Exception(
                'Can not create user pod %s already exists & could not be deleted' % self.pod_name)
```

Reply:
ahh got ya, I can add the retry logic.